### ISSUE TITLE
Sync game to audio

### ISSUE THREAD

**Issue by Krusher97** [m64p/mupen64plus-GLideN64#20]
Is there any way to sync game to audio? I keep getting these popping sounds which are getting on my nerves.

**Comment by loganmc10** [comment]
Can you give me an example? Is there some easy way I can reproduce this?

To answer your question, no, there is not currently a way to sync to audio

**Comment by Krusher97** [comment]
webm : https://webmshare.com/xOPQn
Save file :[Mischief Makers (U) [!].zip](https://github.com/m64p/mupen64plus-GLideN64/files/1356056/Mischief.Makers.U.zip)


I suppose it could be related to performance but there are times when even GlideN64 does this while showing 100% speed.

**Comment by Krusher97** [comment]
Should I open this issue in the main Mupen64plus github itself? I've been encountering this popping in almost every game now even with GlideN64.

**Comment by loganmc10** [comment]
I'm very interested in solving this, but my issue is that I can't reproduce it. I've heard a few other people talk about popping audio, so I would like to figure out what's up. It could be something up with the speed limiter that mupen64plus uses.

I tested the save game you posted, I played through the ending and it sounded normal on my laptop. I need some more info, how constant is the popping? I would expect some popping the first time you play a game with GLideN64, since it compiles shaders the first time you play. This is similar to "shader stuttering" you might read about with Dolphin users.

**Comment by Krusher97** [comment]
Which operating system are you using? I am running it on Windows 7.

**Comment by loganmc10** [comment]
I've tested it on a laptop with Windows 10 and Linux

**Comment by Krusher97** [comment]
Do you get no popping even in Angrylion during the ending cutscene?

Here is a webm of relatively clear popping during the flute playing : https://my.mixtape.moe/yrentc.webm
The game is Zool/Zoor with GlideN64. This is with the Vsync disabled. Enabling Vsync makes the game audio pop/crackle almost consistently.

**Comment by loganmc10** [comment]
The audio sounds terrible if I use Angrylion, but that is because it's not running at full speed. There is not much I can do about that. Do you have audio issues with non-2D games like Super Mario or Smash Bros?

I have vsync off on my laptop, I did notice a tiny bit of crackle in Mischief Makers with vsync enabled. Are things generally better with vsync off?

**Comment by Krusher97** [comment]
Things are generally better with Vsync off yeah. It doesn't remove the crackling completely though and there is screen tearing.

Is there any way to know what speed the game is running at with Angrylion? I just tried Super Mario 64 and the game itself seems to be running at normal/full speed but the audio crackles like crazy.
https://my.mixtape.moe/lsgean.webm

**Comment by loganmc10** [comment]
> Is there any way to know what speed the game is running at with Angrylion? I just tried Super Mario 64 and the game itself seems to be running at normal/full speed but the audio crackles like crazy.

Not with mupen64plus, GLideN64 implemented its own performance counter and OSD (on screen display), without that there isn't a built-in way to see the VI/s or FPS that I know of.

**Comment by loganmc10** [comment]
Is it better in PJ64? I don't mean with Angrylion because there are some differences there, but do you have the same issue in Mischief Makers using GLideN64 in PJ64?

**Comment by Krusher97** [comment]
PJ64 has an option to sync game with audio but there are other problems with PJ64's audio like a constant high pitched note behind almost every music track.

**Comment by Krusher97** [comment]
Just noticed your edit about the webm links. Any other site I can use to share them?

**Comment by loganmc10** [comment]
Dropbox or Google Drive or something would work.

> To elaborate a bit more, the Sync game to audio setting eliminates crackling but its replaced by the game getting little pauses in the middle so I guess it doesn't solve the problem completely.

Yeah this is basically the problem with Sync to Audio settings. It just replaces audio issues with full-game stuttering.

Can you check something for me? Can you download this tool:

https://vvvv.org/contribution/windows-system-timer-tool

And tell me what your current value is? (the item in bold)

**Comment by Krusher97** [comment]
![timer tool](https://user-images.githubusercontent.com/25664221/31405180-e9a6ab52-ae1b-11e7-87de-38bf121d4372.png)

Mario 64 with angrylion: https://drive.google.com/open?id=0B1sugCZh1a3ZTE84WU03enpLc2M
Zool with GlideN64: https://drive.google.com/open?id=0B1sugCZh1a3ZMnExT05wR3FjQzg

**Comment by loganmc10** [comment]
I hear the audio issues in those clips, but like I said, I don't seem to have the issue.

I'm going to spend some time getting this all to compile using Visual Studio, right now it's compiled using MinGW, perhaps a switch to VS will produce a better product, we'll see

**Comment by Krusher97** [comment]
I noticed that [this](http://mupen64plus.org/wiki/index.php?title=Mupen64Plus_Plugin_Parameters) wiki lists buffer size as a customizable parameter. Is there anyway to modify those in the GUI?

**Comment by loganmc10** [comment]
This package actually uses a different audio plugin (https://github.com/m64p/mupen64plus-audio-sdl2), I added a couple options to configure the buffer/latency: They'll be under the Audio Plugin section in the GUI if you download the latest from the website:

BUFFER_SIZE: Defaults to 1024, needs to be a power of two. You can try increasing or decreasing this, it may help/solve your problem.

MAX_LATENCY: Defaults to 300. This represents the max latency (in ms) that the audio plugin will allow. If the audio latency becomes higher than 300ms, it will skip some audio to keep the latency within the acceptable range.

**Comment by Krusher97** [comment]
Played around with the setting for a while with no results.
The only thing that reduced the crackling somewhat was changing the Default frequency to 44100 from 33600 but it didn't completely fix it. What is the default frequency based on?

**Comment by loganmc10** [comment]
The default frequency is only used if the game doesn't set the frequency, which is almost never. The only game I know of that needs that is Ocarina of Time: Master Quest, which was a ROM ripped from a GameCube disc/emulator. Every actual retail N64 game properly sets the frequency, and the default frequency is never used. You could change that to 0 or any other value and it shouldn't make any difference.

**Comment by Krusher97** [comment]
I guess the reduced crackling was just placebo then. Just ran it a few more times with it set to 0. Different results each time.

**Comment by loganmc10** [comment]
I haven't given up on this, I've made an attempt at fixing this problem, can you try the newest build on the site: https://m64p.github.io

Let me know if it helps at all, if it does, but not 100%, we can potentially tweak some of the values being used

**Comment by Krusher97** [comment]
Just tried it. Couldn't notice a difference.

**Comment by KingOfHyrule** [comment]
For the record, I don't have any crackling sound. Are you sure you're running at full speed? I didn't tested many games, but the only time I ever had crackling was when I was running near full speed (58~VI/s), and that was because my video card was running at power saving mode, since the load was too small.

What are your specs?

**Comment by Krusher97** [comment]
i5 6500, 750Ti, 8GB RAM.
I don't think its my computer because I can run other emulators without crackling. Dolphin used to have the same issue for me until they added the cubeb backend at which point it stopped having crackling across all backends for me.
In Mupen64plus I get crackling whenever I enable Vsync even in Mario 64 using GlideN64 in screens where I can get 500% speed without Vsync and speed limiter.

**Comment by KingOfHyrule** [comment]
Indeed, your PC is more than enough.

It may sound like a dumb suggestion, but did you make sure to set your NVIDIA control panel to run your card at maximum performance with mupen? I mean, it may be the case that your card is throttling due to the low usage, and thus performance gets inconsistent.

**Comment by Krusher97** [comment]
Its set to maximum performance by default due to the global settings.

**Comment by KingOfHyrule** [comment]
Okay, so I noticed that this also happens to me, every now and then. One particular setting that I use is cxd RSP, since it outputs better sound than regular mupen HLE RSP. The rest is set to default values, except 4x IR and bilinear filter in GlideN64 (probably irrelevant?).

Most common place where it happens is when you try to play the ocarina in zelda OOT. Sometimes everything plays fine, sometimes it pops.
Changing pages fast in the subscreen also seems to trigger it, but is a lot less common.

I've had sessions where the pop never showed up, and also had sessions where it showed up almost immediately. So it's a very weird problem, maybe related to timing getting messed up?

Also, angrylion doesn't seems to have this problem, but I haven't used it enough to be sure.

**Comment by loganmc10** [comment]
> Also, angrylion doesn't seems to have this problem, but I haven't used it enough to be sure.

GLideN64 uses OpenGL a lot more heavily than angrylion. If you've ever used or read about the Dolphin emulator, you may have come across "shader stutter". GLideN64 suffers from the same problem. When a new scene is being rendered, a new shader may need to be compiled, which will cause a pause in the emulator, which messes up the audio.

GLideN64 recently implemented a feature where it will save the shaders it uses, so the next time you play the game, it will actually load all the shaders when the game starts, not during gameplay. This should help with some of these issues. But the first time you play a game you may still come across it, since it needs to build up that cache.

Anyway that is one explanation, audio issues are typically performance related (like the shader stutter), or there are some games that do get out of sync since the emulator is not cycle accurate.

**Comment by KingOfHyrule** [comment]
Yes, I'm familiar with shader stutter from Dolphin. So, that is also a problem in GlideN64...

Well, I haven't tested this particular scenario - stutter/audio pop after shader cache is built - but can do it later today. Is there any particular way to test this consistently? Where does the shader cache stays stored?

Regarding cycle accuracy, I guess the only way to do this is to sync the audio, or use stretching/skipping right? Does mupen's audio plugin support that?

**Comment by loganmc10** [comment]
> Where does the shader cache stays stored?

On Windows -> C:\Users\you\AppData\Roaming\Mupen64Plus\shaders

> Is there any particular way to test this consistently?

Well the first time you play a game it should create a .keys file and a .shaders file for the game. One game I really notice this is GoldenEye, the intro to the Dam level, it usually skips pretty bad the first time you play, but the 2nd+ time, it shouldn't skip/stall between cutscenes.

> or use stretching/skipping right? Does mupen's audio plugin support that?

The audio plugin that the Android version (mupen64plus AE) uses supports stretching, but not the audio plugin used on PC. It's possible to implement, it just hasn't been done

**Comment by KingOfHyrule** [comment]
Sorry for the delay, haven't had any spare time to test this until now.

I actually don't have GoldenEye, but I tried testing OOT and had some results:

Without shaders there is stutter, but only on the subscreen (pause menu). I tried booting the game three times and playing songs for over a minute each time, but had no pop in the sound.

I put my shader cache back on the folder, and started playing normally. After some time, the sound starts to pop every now and then, suggesting that it indeed is a timing problem, not related to shader stutter.

All tests done with cxd RSP.

**Comment by loganmc10** [comment]
> suggesting that it indeed is a timing problem, not related to shader stutter.

It's possible, but it could still be performance related, just unrelated to shader stutter (for instance, your anti-virus kicks in and causes the game to have a hiccup, some scene takes a long time to render for whatever reason, etc...)

**Comment by KingOfHyrule** [comment]
I see. Well, my PC has very little (almost nothing) running in the background, and my CPU is far from being stressed, so I really think this shouldn't be happening.

**Comment by loganmc10** [comment]
I agree, I've tried a few things to minimize the issue, but adding some kind of sound stretching feature to the audio plugin (like is done on Android), is probably the best solution

**Comment by Tasosgemah** [comment]
Sound stretching is something N64 audio plugins needed for decades. Popping/cracking was always an issue with many setups and combinations. I also get some popping in some games but i just accepted it because when it comes to N64 emulation it was always a thing. But if there is a way for this to be fixed it will be a huge deal for me.

Question is, will sound stretching add too much audio latency? There is a similar option in Citra but it adds so much latency (like half a second) that it's better to just deal with the pops.

**Comment by loganmc10** [comment]
The Android version of mupen64plus has sound stretching. I don't really notice the latency, it's certainly not half a second, but @fzurita might have some actual numbers, he's the one that wrote it.

**Comment by fzurita** [comment]
To be truthful, I haven't tried to make any measurements, by from my non-scientific testing, it feels like 50-100ms.

**Comment by fzurita** [comment]
If you wanted to take the Android audio stretching, it wouldn't be hard, all you need is the Sound touch library. Although, Windows may not have pthread.

**Comment by fzurita** [comment]
@loganmc10 I created this pull request to add audio time stretching to the audio plugin:

https://github.com/m64p/mupen64plus-audio-sdl2/pull/3

**Comment by fzurita** [comment]
If anyone wants to try the time stretching, here is a compiled version: https://drive.google.com/file/d/1MAGd6KJtE4cJxsBmbjntIdMBLe7yZB2A/view?usp=sharing

See https://github.com/m64p/mupen64plus-audio-sdl2/pull/3 for utilization instructions.

**Comment by Tasosgemah** [comment]
I tested Top Gear Rally (which is one of the hard games to emulate in terms of sound) and it sounds worse with time stretching ON. With it OFF it sounds good but it skips/glitches once every one minute or so.

**Comment by fzurita** [comment]
Ok, thanks for testing. Did you feel any audio latency? 

For Top Gear Rally, can you try setting these parameters?

SECONDARY_BUFFER_NBR = 20
SAMPLING_RATE = 48000

**Comment by Tasosgemah** [comment]
SECONDARY_BUFFER_NBR was already "20"

Sampling rate was "0" so i changed that to 48000.

Now Top Gear Rally sounds good but the skip/glitch issue every 1 minute is still there. In addition there is a very noticeable sound lag. It's more than half of a second.

And yeah, Top Gear Rally always had trouble with sound in all emulators until recently. Now at least it's good if you don't pay attention to the occasional skips and glitches. It was horrible before.

**Comment by fzurita** [comment]
Ok, when using 48000, I suspect someone down the line is doing re-sampling, which is adding delay. Can you try various frequencies like 32000, 44100, 96000, etc to see if the delay is still there? We want to find at what frequencies re-sampling doesn't happen.

Also, with Top Gear Rally, those sound issues may be coming from the game itself. As in the game is producing those skips and glitches.

**Comment by Tasosgemah** [comment]
I have the real game, the skips/glitches aren't there. It's just that this game has some weird sound processing, BOSS games always have something hard to emulate.

I'm gonna test the other frequencies later on, thanks for the responses.

**Comment by fzurita** [comment]
I'm aware that they are not in the real N64, I was trying to say that it was an emulation bug producing those skips/glitches.

**Comment by Tasosgemah** [comment]
Ah ok got confused, English is not my native language.

**Comment by Pixelnarium** [comment]
At first I had to figure out that this build is for the 32bit version... ;-) Normally I am using Linux. This does not solve the problem. I thought that timestreching is for systems that are not powerful enough. The game itself runs full speed -> 60VI/s. I tried it on a i5-4660 with 3.2Ghz and a GTX 1060.

With the 32bit build with your plugin the problem is even worse. The audio desyncs right in the first few seconds.

**Comment by loganmc10** [comment]
A good way to test this is to pause the game (p key) when the audio is desynced. When you unpause it, is the audio fixed? Or is it still desynced?

If it's still desynced, then it's not the audio plugins fault, the emulator has screwed up the audio

**Comment by Pixelnarium** [comment]
When I pause the game it does not fix the issue. So should I report it to mupen64plus-core? Using rsp-cxd4 with angrylion does not change anything.

**Comment by loganmc10** [comment]
You can try changing the "CountPerOp" setting for mupen64plus-core to 1. This is just an emulator timing issue, many cutscenes have issues like this currently

**Comment by Pixelnarium** [comment]
This does not help. But if it is a known problem it "Okay" in terms of reporting bugs.

**Comment by Pixelnarium** [comment]
Just wanted to add some informations for "completeness" and future reference. I don't have a US Version of the N64 only a PAL one. So I don't know if this happens on a US console too:

Running the PAL version of the game: The intro is lip synched. Running the US version of the game (again on PAL hardware through Everdrive at normal 60Hz and not 50) the intro desyncs too. Not "that bad" as in the emulator but noticeable.

So it seems that the game itself does not care about synced sound that much.

**Comment by pedrib** [comment]
I know this is an old topic, but I have been getting a lot of audio cracking, and after turning on the FPS counter, I'm pretty sure this is a performance issue. Once there is a major drop in FPS below 30, the audio starts cracking. 

This only happens in some games (the heavier ones like Indian Jones, Rogue Squadron, etc) and mostly in fullscreen mode, when I reduce to a window with 640x480 it seems to solve it. However I have set my resolution to 640x480 in the settings, shouldn't this be respected in full screen? If I turn on the internal and external resolution, I notice that the external resolution is set to the fullscreen size instead of 640x480.

My machine is a i7-7500U laptop with Intel 620 graphics and 16 GB of RAM. I guess I need more power?

**Comment by loganmc10** [comment]
With GLideN64, go into the video plug-in settings and look for "UseNativeResFactor" and set it to 1 or 2, that will force the rendering resolution to stay at a factor of 1x or 2x the games native resolution, and it will greatly improve performance

**Comment by pedrib** [comment]
@loganmc10 thanks for the tips. That indeed solves the performance issues, although now it doesn't look as good!
I will experiment with different factors until I reach a good compromise. So far 3x looks good for my setup.

So this is indeed a performance problem. I guess I'll have to get something beefier in the future.

And thanks for making this awesome project and putting everything together for ease of use.
